File: 0494-target-sum/0494-target-sum.java
// Memoization
class Solution {
    public void numTarSumWays(int idx, int[] nums, int target, int[] count, int[][] dp) {
        if (target < -1000 || target > 1000) return;
        
        if(idx < 0) {
            if(target == 0) count[0]++;
            return ;
        }
        if(dp[idx][target + 1000] != -1) {
            count[0] += dp[idx][target + 1000];
            return ;
        }

        int beforeCount = count[0];
        numTarSumWays(idx-1, nums, target - nums[idx], count, dp);
        numTarSumWays(idx-1, nums, target + nums[idx], count, dp);
        int ways = count[0] - beforeCount;

        dp[idx][target + 1000] = ways;
    }

    public int findTargetSumWays(int[] nums, int target) {
        int n = nums.length;
        int[] count = new int[1];
        int[][] dp = new int[n][2001];
        for(int i=0; i<n; i++) {
            Arrays.fill(dp[i], -1);
        }
        numTarSumWays(n-1, nums, target, count, dp);
        return count[0];
    }
}

// class Solution {
//     public void numTarSumWays(int idx, int[] nums, int target, int[] count, Integer[][] dp) {
//         if (idx < 0) {
//             if (target == 0) count[0]++;
//             return;
//         }

//         // Memoization check
//         if (dp[idx][target + 1000] != null) {
//             count[0] += dp[idx][target + 1000]; // reuse cached result
//             return;
//         }

//         int before = count[0];
//         numTarSumWays(idx - 1, nums, target - nums[idx], count, dp);
//         numTarSumWays(idx - 1, nums, target + nums[idx], count, dp);
//         int ways = count[0] - before;

//         dp[idx][target + 1000] = ways; // store result
//     }

//     public int findTargetSumWays(int[] nums, int target) {
//         int n = nums.length;
//         int[] count = new int[1];
//         Integer[][] dp = new Integer[n][2001]; // memo table

//         numTarSumWays(n - 1, nums, target, count, dp);
//         return count[0];
//     }
// }

File: 0494-target-sum/0494-target-sum.py
# class Solution:
#     def findTargetSumWays(self, nums: List[int], target: int) -> int:
#         n = len(nums)
#         totalSum = sum(nums)
#         if (totalSum - target) % 2 != 0:
#             return 0
#         if target > totalSum:
#             return 0
#         s2 = (totalSum - target) // 2
#         dp = [[0 for _ in range(s2 + 1)] for _ in range(n)]

#         # base cases
#         for i in range(0, n):
#             dp[i][0] = 1
#         if nums[0] <= s2:
#             dp[0][nums[0]] = 1

#         # DP for loops
#         for i in range(1, n):
#             for j in range(1, s2 + 1):
#                 take = 0
#                 if j >= nums[i]:
#                     take = dp[i-1][j - nums[i]]
#                 notTake = dp[i-1][j]
#                 dp[i][j] = take + notTake
        
#         return dp[n-1][s2]

class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        n = len(nums)
        totalSum = sum(nums)

        if totalSum - target < 0 or (totalSum - target) % 2 != 0:
            return 0

        s2 = (totalSum - target) // 2

        dp = [[0 for _ in range(s2 + 1)] for _ in range(n)]

        # base case
        if nums[0] == 0:
            dp[0][0] = 2
        else:
            dp[0][0] = 1

        if nums[0] != 0 and nums[0] <= s2:
            dp[0][nums[0]] = 1

        for i in range(1, n):
            for j in range(0, s2 + 1):
                take = 0
                if nums[i] <= j:
                    take = dp[i - 1][j - nums[i]]

                notTake = dp[i - 1][j]

                dp[i][j] = take + notTake

        return dp[n - 1][s2]